Keybindings rework3

kugel- opened the pull request:
New API functions for keybindings that have a userdata pointer. This is needed to allow for potential non-C plugins to store context information (e.g. this pointer for object instances).

b4n commented:
Apart possibly the names, LGTM.

Any other opinions on that?

BTW, if we don't want plugins to set default keybindings, we might want to take the opportunity to remove the `key` and `mod` parameter, possibly moving them to a `_full()` variant if necessary.

b4n commented:
@codebrainz @frlan @elextr @eht16 opinions on the API?

codebrainz commented:
@b4n I still like [my similar change](e3e299be81191ebedcabbd5ac8f819544a5f34cd) better :)

The only thing I can see that might be a problem is not having a destruction notifier (or having a `_full` variant of the function). This would be useful for bindings like GeanyPy.

elextr commented:
@codebrainz your similar change comes up as a 404 for me?

@kugel- this is mildly out of scope of this change, so its probably another PR, but plugin keybindings should not override existing ones, they should get an error back if they try to.

kugel- commented:
I don't mind a destroy notify although I don't think it's necessary for bindings (but maybe it can help). Thinking about it, might be useful for vala plug-ins to allow for passing lambda expressions

Am 27. Mai 2015 01:21:17 MESZ, schrieb Matthew Brush <notifications@github.com>:
>@b4n I still like [my similar
>change](e3e299be81191ebedcabbd5ac8f819544a5f34cd) better :)
>
>The only thing I can see that might be a problem is not having a
>destruction notifier (or having a `_full` variant of the function).
>This would be useful for bindings like GeanyPy.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/geany/geany/pull/376#issuecomment-105695792

codebrainz commented:
If it's not generally useful, it could use the pattern of GLib and have a `_full()` version of the function for bindings that allows a destruction notifier, so that in the common case plugins don't have to pass an extra `NULL` all the time.

Also as @b4n said, we should probably drop the key and mod parameters since your not supposed to use them anyway (see API docs), they just stuck around for compatibility (or laziness to write a new function or whatever). IMO, we shouldn't move those two params to a `_full()` version as that confuses with the purpose of `_full()` functions with how GLib uses them, for bindings to pass a destroy notifier.

b4n commented:
I think I like @codebrainz's name better (e.g. `GeanyKeyBindingCallbak`), and agree the destroy notify could be useful especially for bindings (though in most cases if it's done by humans they can just destroy it when unloading the plugin, it's virtually the same).

But I like the fact this one also adds it to keybinding group callbacks (although I don't really like those), and that it adds a boolean return value allowing the per-keybinding callbacks to have all the features of the group ones.
This said, I don't know how sensible it is to allow the same keybinding on different handlers in practice, for most things it will just not work well -- but well, we do allow that currently in the keybindings settings, so I guess improving support for that makes sense.

re. default value for keybindings, agreed.  if we want to keep them, they should at least only be optional and be disabled if they conflict with an existing one or something like that.  Which basically means it's not reliable and then mostly useless.

b4n commented on src/keybindings.c:
```diff
@@ -172,12 +173,47 @@ GeanyKeyBinding *keybindings_set_item(GeanyKeyGroup *group, gsize key_id,
+ * @since 1.25
+ * @see See plugin_set_key_group_with_handler
+ **/
+GeanyKeyBinding *keybindings_add_item_with_handler(GeanyKeyGroup *group, gsize key_id,
```
just noticed it's "*add*_item", which is different from what we already have, and a bit misleading as it actually doesn't add anything -- it sets it

b4n commented:
Hum, what about something like `keybindings_group_set_item()` or alike, as it works on a `GeanyKeyGroup`?  Most other API are prefixes with the structure name (stripping `Geany` prefix).

kugel- commented:
So I uploaded another version

- added GDestroyNotify (as a seperate commit actually)
- renamed *Handler to *Hook
- renamed *_with_handler to *_full
- renamed parameter names
- make use of GPtrArray free_func for calling the destroy notify

kugel- commented:
Fixed a invalid free (was called for builtin groups) and replaced a too-recent glib functions with 2.22 compatible code.

b4n commented:
about names, what about:

1. Use `_full()` (although it's a bit misleading as an argument would actually be different, but I have nothing better to suggest)
2. Use `GeanyKeyBindingCallback` and `GeanyKeyBindingGroupCallback`.
   1. If you (@kugel-) really care about the name being a "member" (naming-wise) of `GeanyKeyGroup` (which you shouldn't as it wasn't the case with `GeanyKeyCallback` and IIRC the first proposal here), just typedef it to `GeanyKeyBindingGroup`.

b4n commented:
BTW, to summarize some stuff from IRC:

* I don't like `hook` at all, which seem to basically be the only term GLib and friends never use for this.
* I'm under the impression that `Func` is more generally used as a synchronous callback or a "virtual method" than an asynchronous callback.

But actually, I don't care much about the name of the callback type, it's not like anybody really have to see it.  Function name is a lot more important, and I guess `_full()` is fine.

kugel- commented:
Last update only renames from Hook to Func

b4n commented on src/keybindings.c:
```diff
@@ -79,33 +79,33 @@ static gboolean check_current_word(GeanyDocument *doc, gboolean sci_word);
+static void GeanyKeyBindingFunc_switch_tableft(guint key_id);
+static void GeanyKeyBindingFunc_switch_tabright(guint key_id);
+static void GeanyKeyBindingFunc_switch_tablastused(guint key_id);
+static void GeanyKeyBindingFunc_move_tab(guint key_id);
```
WTF

kugel- replied:
That's right. I should have looked more carefully

b4n commented on src/keybindings.c:
```diff
@@ -194,12 +194,67 @@ GeanyKeyBinding *keybindings_set_item(GeanyKeyGroup *group, gsize key_id,
 }
 
 
+/** Creates a new keybinding using a GeanyKeyBindingFunc and attaches is to a keybinding group
```
"[sic] attaches **it** to a [sic]"

b4n commented on src/keybindings.c:
```diff
@@ -194,12 +194,67 @@ GeanyKeyBinding *keybindings_set_item(GeanyKeyGroup *group, gsize key_id,
+ * @param label Label used in the preferences dialog keybindings tab. May contain
+ * underscores - these won't be displayed.
+ * @param menu_item Optional widget to set an accelerator for, or @c NULL.
+ * @return The keybinding - normally this is ignored.
```
params should be in the correct order.
destroy_nofity should be documented

b4n commented on src/keybindings.c:
```diff
@@ -244,22 +302,22 @@ static void init_default_kb(void)
 	ADD_KB_GROUP(GEANY_KEY_GROUP_TOOLS, _("Tools"), NULL);
 	ADD_KB_GROUP(GEANY_KEY_GROUP_HELP, _("Help"), NULL);
-	ADD_KB_GROUP(GEANY_KEY_GROUP_FOCUS, _("Focus"), cb_func_switch_action);
+	ADD_KB_GROUP(GEANY_KEY_GROUP_FOCUS, _("Focus"), GeanyKeyBindingFunc_switch_action);
```
same

b4n commented on src/pluginutils.c:
```diff
@@ -307,6 +309,36 @@ GeanyKeyGroup *plugin_set_key_group(GeanyPlugin *plugin,
+ * @param section_name Name used in the configuration file, such as @c "html_chars".
+ * @param count Number of keybindings for the group.
+ * @param cb New-style group callback, or @c NULL if you only want individual keybinding callbacks.
+ * @param pdata Plugin specific data, passed to the group callback.
```
destroy_notify is missing

kugel- commented:
I've fixed the issues you outlined. I hope it's ready to go now.

b4n commented on src/keybindings.h:
```diff
@@ -37,12 +37,33 @@ G_BEGIN_DECLS
+ * You should return @c TRUE to indicate handling the callback. (Occasionally, if the keybinding
+ * cannot apply in the current situation, it is useful to return @c FALSE to allow a later keybinding
+ * with the same key combination to handle it). */
+typedef gboolean (*GeanyKeyGroupFunc)(GeanyKeyGroup *group, guint key_id, gpointer pdata);
```
missing `@since`

kugel- replied:
Argh, I double checked, but only at the functions :'( Will fix in a few minutes

b4n commented on src/keybindings.h:
```diff
@@ -37,12 +37,33 @@ G_BEGIN_DECLS
+ * You should return @c TRUE to indicate handling the callback. (Occasionally, if the keybinding
+ * cannot apply in the current situation, it is useful to return @c FALSE to allow a later keybinding
+ * with the same key combination to handle it). */
+typedef gboolean (*GeanyKeyBindingFunc)(GeanyKeyBinding *key, guint key_id, gpointer pdata);
```
same

b4n commented:
LGTM, apart API bump (which should probably be done at merge time as this and #469 will conflict)

kugel- commented:
@since added

b4n commented on src/keybindings.h:
```diff
@@ -37,12 +37,42 @@ G_BEGIN_DECLS
+ * with the same key combination to handle it).
+ *
+ * @since 1.25. */
+ typedef gboolean (*GeanyKeyGroupFunc)(GeanyKeyGroup *group, guint key_id, gpointer pdata);
```
spurious space at the start of the line (please dont kill me :])

kugel- commented:
For the record (as mentioned on irc), i fixed the whitespace. Hopefully good now

b4n commented on src/keybindings.c:
```diff
@@ -194,12 +194,68 @@ GeanyKeyBinding *keybindings_set_item(GeanyKeyGroup *group, gsize key_id,
+	/* For now, this is intended for plugins only */
+	g_assert(group->plugin);
+
+	kb = keybindings_set_item(group, key_id, NULL, 0, 0, kf_name, label, menu_item);
```
hum, the function does have `key` and `mod` parameters, but they aren't propagated

kugel- replied:
Am 18. Juni 2015 01:56:55 MESZ, schrieb Colomban Wendling <notifications@github.com>:
>> + *
>> + * @since 1.25
>> + * @see See plugin_set_key_group_full
>> + **/
>> +GEANY_API_SYMBOL
>> +GeanyKeyBinding *keybindings_set_item_full(GeanyKeyGroup *group,
>gsize key_id,
>> +		guint key, GdkModifierType mod, const gchar *kf_name, const gchar
>*label,
>> +		GtkWidget *menu_item, GeanyKeyBindingFunc cb, gpointer pdata,
>> +		GDestroyNotify destroy_notify)
>> +{
>> +	GeanyKeyBinding *kb;
>> +
>> +	/* For now, this is intended for plugins only */
>> +	g_assert(group->plugin);
>> +
>> +	kb = keybindings_set_item(group, key_id, NULL, 0, 0, kf_name,
>label, menu_item);
>
>hum, the function does have `key` and `mod` parameters, but they aren't
>propagated
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/geany/geany/pull/376/files#r32688415

Oops good catch, clearly a bug.

b4n commented:
Also, I don't know where the "should we keep `key` and `mods` in the new API" went, but I would have swore it was here…?  anyway, I see benefit of both, but keeping them might mean we need to actually do something at some point to avoid conflicts (ofc it's out of scope here); so I don't mind either way.

codebrainz commented:
We should reach a consensus on whether it is recommended for plugins to set key/mod parameters. If the consensus is that they shouldn't use them (as implied by the warning in the doxygen documentation for the existing function), then we should not be adding them to any new API. If the consensus is that it's fine to use them, then we should add them to the new function and remove the warning in the existing function's doxygen documentation.

codebrainz commented:
Or, if we want to rush this through in time for next release, we could just entirely drop the Doxygen comment so it's not officially part of the API yet. Or barring that, we could do something like this:

```c
/**
 * ...
 * @param reserved1 Do not use, always pass 0.
 * @param reserved2 Do not use, always pass 0.
 * ...
 */
GeanyKeyBinding *keybindings_set_item_full(GeanyKeyGroup *group, gsize key_id,
		guint reserved1, gint reserved2, const gchar *kf_name, const gchar *label,
		GtkWidget *menu_item, GeanyKeyBindingFunc cb, gpointer pdata,
		GDestroyNotify destroy_notify);
```

codebrainz commented:
FWIW, my personal opinion, as a plugin developer, is that it's useful to be able to set default key/mods. The only change I would make is to pass a string like `"<ctrl><shift>f"` rather than two ints like `GDK_s, GDK_SHIFT_MASK | GDK_CONTROL_MASK`. I wouldn't even mention that here now except that we're stuck with whatever this PR results in for a long time, and it would suck having to add a 3rd function to do same thing.

elextr commented:
I'm not happy to be rushing to get this in to an already way overdue release when there is still disagreement over things like setting defaults for keybindings.

I am very much against plugins setting default keyvalues because of the inevitable conflicts and as @codebrainz says we are likely to be stuck with this for some time.

As @b4n said on IRC, since the existing plugin interface is not going away they do not need to change, so they won't lose the defaults that users have sadly become used to.  And new plugins using this API should not set keyvalues, so the arguments should be removed.

kugel- commented:
My opinion is that we should keep them since they are useful. We should replace our recommended "do not use" by "use with care". And we should probably implement a UI to resolve conflicts on activation. 

To me the fear of conflicts is just FUD spread by @elextr. We haven't actually seen in practice that plugins abuse this functionality nor have people report problematic plug-ins. So not keeping the keybindings would be removing a useful ability for plug-ins purely based on FUD. And we'd be removing it because we're unwilling to handle conflicts in a user friendly way even if it's clearly possibly and not even too hard.


At last, we can always decide to just ignore the defaults if they become problematic. It's telling that we haven't even considered this for the existing API, because in practice there is actual problem.

b4n commented:
@frlan As you were the one adding the note on `keybindings_set_item()` in 304f3a8b476351230bc0324597f0689e9cddebeb, what is your opinion on this? (and reasons for initial change maybe)

Also, I just tried and it seems that Geany's keybindings win over the plugin's ones (just tried to make a plugin bind something to `<primary>o` by default, and that still does the Geany thing and not the plugin's).  I'm not sure what happens if we change one of Geany's keybindings (or add a new one) that conflicts with a plugin's one, though, but I'm not sure how we deal with this regarding user kbs (I think we never change an existing config's keybindings, but I'm not totally sure).

> The only change I would make is to pass a string like `"<ctrl><shift>f"` rather than two ints like `GDK_s, GDK_SHIFT_MASK | GDK_CONTROL_MASK`.  […]

This might make some sense, as most generally this is a static value, and it has definitely the benefit of shortness (and no need to include `<gdk/gdkkeysyms>` or worry about the different key names on various GDK versions -- although this last one is mostly moot as we now depend on GTK 2.24 which has the new ones).  Also it's slightly easier to use `<primary>` than thinking of (the new) `GEANY_PRIMARY_MOD_MASK`.
OTOH, it's easier to introduce a typo (but well, admittedly if the developer doesn't realize it's questionable whether the default matters :)

elextr commented:
> Also, I just tried and it seems that Geany's keybindings win over the plugin's ones (just tried to make a plugin bind something to <primary>o by default, and that still does the Geany thing and not the plugin's). 

Had a look at the code.  Geany wins because set keybindings functions do not remove the existing binding AFAICT, so who wins depends on the search order, which is the order things appear in the keybindings dialog I think.

kugel- commented:
Makes sense, plugin key groups are appended to the array, and our dispatcher runs in fifo order. So yea, it seems it has always been the case that plugins can't practically override Geany keybindings.

I would also like to put a reminder that conflicting keybindings aren't necessarily a bad thing: because we can chain key bindings, plugins can provide context/widget-specific keybindings and let Geany handle other situations by returning FALSE in the handler. This is a legitimate use case IMO.

@b4n, @codebrainz  it *seems* nice on the first look, but has the big disadvantage of preventing the compiler to do compile time checks for typos. I don't feel strong about this, though.

elextr commented:
@kugel- how do you provide widget specific keybindings?

kugel- commented:
Install a normal keybinding and check for the focused widget in the callback.

The normal GTK way would be to install a handler for the widgets key-press-event signal but I'm not sure this method works well with how we configure keybindings. But I think geanywebhelper does this (without installing defaults) and uses questionable hacks to look up the configuration for its keybinding inside the handler.

b4n commented:
> But I think geanywebhelper does this (without installing defaults) and uses questionable hacks to look up the configuration for its keybinding inside the handler.

It only does on its separate window (if configured to have one) so it can react to keybindings received there (as Geany will only receive them on its own window).

b4n commented:
> Install a normal keybinding and check for the focused widget in the callback.

"normal" is not really correct here, as Geany will stop searching after the first match for item-specific handlers (that return nothing, remember).  Only key group callbacks can return whether they handled the event or not, and if not Geany will keep on matching keybindings.

With your version here however, item-specific callbacks also return whether they handled the event or not, so can do this too.

elextr commented:
> Install a normal keybinding and check for the focused widget in the callback.

Ahhh, focussed widget, ok, was wondering since the widget is not passed to the kb callback.

> The normal GTK way would be to install a handler for the widgets key-press-event signal but I'm not sure this method works well with how we configure keybindings. But I think geanywebhelper does this (without installing defaults) and uses questionable hacks to look up the configuration for its keybinding inside the handler.

Yeah, I think Geany's keyevent catches all keys first, then, any it doesn't know about are returned to GTK to send to the widgets IIRC so the GTK widget handler should work if that key is not defined in Geany.

kugel- commented:
Fixed passing the default keybindings

kugel- commented:
@b4n @codebrainz ping :-)

kugel- commented:
@b4n @codebrainz ping :-)

kugel- commented:
I only updated the @since tags and added a commit to bump API version. Please merge

b4n commented:
LGTM.

Only possible question that remains is whether we want to use string key descriptions as suggested by @codebrainz instead of integers.  Then the question is whether the name `_full()` is appropriate when the params get so different.
We could also merge like this and possibly add simpler API later if we want, though it might be a bit too much just for this.

kugel- commented:
I don't necessarily agree that the strings are better. I'm generally in favor of compile time checks where possible

kugel- commented:
Please merge if it looks good :-)

codebrainz commented:
> I don't necessarily agree that the strings are better. I'm generally in favor of compile time checks where possible

I am too, in general, but in this case it's more simple and natural to use strings, and the only way a typo could slip through unnoticed is if the plugin developer doesn't test, and even if they didn't, I think it would simply result in the default not being set, rather than crashing or UB or bad stuff.

What's more, the compiler isn't really doing much checks here, since the type of the key value parameter is not strong, and you could stick just about any valid identifier representing a number in there (ex. `TRUE`, `GDK_CONTROL_MASK`, `G_KEYFILE_KEEP_COMMENTS`, etc.)

b4n commented:
@codebrainz unless you have a strong opinion on this (and would like to provide the tiny patch? :)) I think I'm going to merge this in the next few days as is, which also has the advantage of being more consistent with the current API (and thus the `_full` suffix).

kugel- commented:
@b4n @codebrainz no reaction. merge? :-)

b4n commented on src/keybindings.h:
```diff
@@ -37,12 +37,42 @@ G_BEGIN_DECLS
+ * with the same key combination to handle it).
+ *
+ * @since 1.26 (API 226) */
+typedef gboolean (*GeanyKeyBindingFunc)(GeanyKeyBinding *key, guint key_id, gpointer pdata);
```
I was trying to port some plugins to use this API and realized `key_id` is kinda redundant here with `key->id`.  What about instead making `GeanyKeyBinding::id` public (add a doc comment) and dropping the `key_id` parameter here?
Or do we think it's good because it's more consistent with the group callback?

kugel- commented:
Err, you said you'd be merging this as-is :( But tbh, I don't care about key_id, whatever you prefer is fine with me.

A few months ago I wanted to have GeanyKeyBinding being opaque, but that was before I realized we'll never be able to covert to GtkAction/GAction.

codebrainz commented:
Meh, no super strong feelings, though I do think it's weird to have the closure stuff squished in the middle of the arguments list rather than more typical GLib-style `_full()` functions where the function and data are at the end of the arguments list. In any case, maybe later, if I get sufficiently motivated, I can provide some kind of "wrappers" to smooth over the (perceived) warts, either in my local plugin code or in addition to the API proper.

b4n commented:
> […] though I do think it's weird to have the closure stuff squished in the middle of the arguments list rather than more typical GLib-style `_full()` functions where the function and data are at the end of the arguments list. 

How so?  It's at the end, isn't it?:

```C
GeanyKeyBinding *keybindings_set_item_full(GeanyKeyGroup *group, gsize key_id,
		guint key, GdkModifierType mod, const gchar *kf_name, const gchar *label,
		GtkWidget *menu_item, GeanyKeyBindingFunc func, gpointer pdata,
		GDestroyNotify destroy_notify);

GeanyKeyGroup *plugin_set_key_group_full(GeanyPlugin *plugin,
		const gchar *section_name, gsize count,
		GeanyKeyGroupFunc cb, gpointer pdata, GDestroyNotify destroy_notify);
```

codebrainz commented:
Yeah, I was looking at the commit that added the function, not the later one where that was fixed, my bad.
